### PR TITLE
Test also on PHP 7.2+

### LINKED ISSUE #3185: PHP 7.2 Support?
After updating PHP to version 7.2.0 icingaweb2 does not work anymore

```
Fatal error: Uncaught ErrorException: session_name(): Cannot change session name when session is active in .../icingaweb2/library/Icinga/Web/Session/PhpSession.php:97 Stack trace:
#0 [internal function]: Icinga\Application\ApplicationBootstrap->Icinga\Application\{closure}(2, 'session_name():...', '/usr/share/weba...', 97, Array)
#1 .../icingaweb2/library/Icinga/Web/Session/PhpSession.php(97): session_name('Icingaweb2')
#2 .../icingaweb2/library/Icinga/Web/Session/PhpSession.php(152): Icinga\Web\Session\PhpSession->open()
#3 .../icingaweb2/library/Icinga/Web/Controller/ActionController.php(544): Icinga\Web\Session\PhpSession->write()
#4 .../icingaweb2/library/Icinga/Web/Controller/ActionController.php(489): Icinga\Web\Controller\ActionController->shutdownSession()
#5 .../icingaweb2/library/vendor/Zend/Controller/Action.php(512): Icinga\Web\Controller\ActionController->postDispatch()
#6 .../icingaweb2/library/Icinga/Web in .../icingaweb2/library/vendor/Zend/Controller/Plugin/Broker.php on line 332
```

## Your Environment
```
:~ > icinga2 --version
icinga2 - The Icinga 2 network monitoring daemon (version: r2.6.3-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Arch Linux ARM
  Platform version: Unknown
  Kernel: Linux
  Kernel version: 4.14.3-1-ARCH
  Architecture: armv7l
```

Icingaweb2 version is 2.4.1 on archlinux.

Will there be a php 7.2 support soon?

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
Rationale: Who knows all the upcoming changes to PHP and our obstacles?

**Comment by dnsmichi** [comment]
Mh if nightlies contain bugs, this will break travis tests on PRs. The benefit of testing the git master against nightlies is then gone. Imho this should be done in a Jenkins job, separate from code commits.

**Comment by Al2Klimov** [comment]
@dnsmichi Well, if even *this* PR fails to be tested... then you're damn right.

**Comment by Thomas-Gelf** [comment]
Hint: add `allow_failures` -> `- php: nightly` to your `matrix`

File: library/Icinga/Application/ApplicationBootstrap.php [hunk 1]
```diff
@@ -531,7 +531,7 @@ protected function loadConfig()
      */
     protected function setupErrorHandling()
     {
-        error_reporting(E_ALL | E_STRICT);
+        error_reporting((E_ALL | E_STRICT) & ~ E_DEPRECATED);
```
Comment (dnsmichi) [inline library/Icinga/Application/ApplicationBootstrap.php]: Mh you should see them imho. This allows to react sooner on possibly removed things, especially if you are really testing the new "hot shit".

Comment (Al2Klimov) [reply library/Icinga/Application/ApplicationBootstrap.php]: Just testing. Don't panic.

Comment (dnsmichi) [reply library/Icinga/Application/ApplicationBootstrap.php]: Then edit the PR's title to include "WIP: " as prefix :)

**Comment by Al2Klimov** [comment]
[It seems](https://travis-ci.org/Icinga/icingaweb2/jobs/311807824#L646) PHP 7.2 has broken our session reopening introduced in efa23ad.

**Comment by Al2Klimov** [comment]
1. The tests fail on PHP 7.2+ because of our session reopening.
2. Our session reopening is required as we use `session_write_close()`.

But why don't we close the session just once – at shutdown?

CC @lippserd @Thomas-Gelf

**Comment by lippserd** [comment]
@Al2Klimov `session_start()` locks the session file until `session_write_close()` is called. We close the session immediately after read/write operations to prevent Ajax requests from blocking. Smart eh? 😆

**Comment by Al2Klimov** [comment]
Gentlemen, it's your turn. B-)

**Comment by Thomas-Gelf** [comment]
Short note: the former session code looks like it tried to avoid issues with "headers already sent". No idea for what use case, but it looks like the new variant would break that countermeasure. Eventually @lippserd can have a look at this. Filter -> good catch! Static code analysis is not that bad at all :laughing:.

**Comment by Al2Klimov** [comment]
@Thomas-Gelf What break? I've tested that stuff w/

```php
class IndexController extends Controller
{
    protected $requiresAuthentication = false;

    public function indexAction()
    {
        $s = Session::getSession();
        $s->write();
        $s->write();
        $s->write();
        var_dump(headers_list());
    }
}
```

**Comment by Thomas-Gelf** [comment]
Nothing is wrong with your code alone, and I'm sure tests are fine. So, it's no break per se, but it changes current behavior. To me the old code looks like it has intentionally been placed there. It told PHP to not send headers on it's own (triggered by session_start) at that early stage. There might be a reason for this, other components might rely on deferred header sending. But you should better discuss this with @lippserd.

**Comment by Al2Klimov** [comment]
... or w/ @nilmerg the author of efa23ad.

**Comment by Thomas-Gelf** [comment]
@Al2Klimov: just in case you're wondering, the referenced issue 5510 refers to dev.icinga.org, it's #457 here on GitHub.

**Comment by Al2Klimov** [comment]
No, I'm not. But thank you anyway.

**Comment by Al2Klimov** [comment]
Just tested w/

```php
class IndexController extends Controller
{
    protected $requiresAuthentication = false;

    public function indexAction()
    {
        $s = Session::getSession();
        $s->write();
        while (ob_get_level()) {
            ob_end_clean();
        }
        echo 'foobar';
        $s->write();
        var_dump(headers_list());
    }
}
```

(It doesn't work.)

**Comment by Al2Klimov** [comment]
... but IMO it doesn't break anything... except write operations at shutdown time. :/

**Comment by Al2Klimov** [comment]
Note:

The latest test fails on my WS for an other error than at Travis. I'll investigate all the things.

**Comment by Al2Klimov** [comment]
Notes:

* It seems https://github.com/PhenX/php-svg-lib have upgraded their license to LGPLv3 – as discussed w/ @lippserd (offline) – never mind: https://www.gnu.org/licenses/gpl-faq.html#AllCompatibility
* 8778a63 will disappear in favor of https://github.com/PhenX/php-svg-lib/pull/25

File: library/Icinga/Chart/Donut.php [hunk 2]
```diff
@@ -344,7 +344,7 @@ protected function encode($content)
 
     protected function renderAttributes(array $attributes)
     {
-        $html = [];
```
Comment (Al2Klimov) [inline library/Icinga/Chart/Donut.php]: @theFeu I'm afraid we can't code like this as long as we don't have green light from @lippserd.

**Comment by Al2Klimov** [comment]
## TODO(@Al2Klimov)

* [ ] Wait for @lippserd to merge #3211
* [ ] Rebase against master
* [ ] Eliminate 3a383d0 and edbf4ae
* [ ] See whether https://travis-ci.org/Icinga/icingaweb2/jobs/316510831#L1219 persists

**Comment by Al2Klimov** [comment]
@lippserd It seems [the only obstacle left](https://travis-ci.org/Icinga/icingaweb2/jobs/316831640#L1218) hasn't gone, but occurs [only in PHP 5.3](https://travis-ci.org/Icinga/icingaweb2/builds/316831633). Seriously, isn't it time to drop PHP 5.3 testing?

> Also, future versions of Icinga Web 2 will drop support for outdated PHP versions.

– [doc](https://github.com/Icinga/icingaweb2/blob/master/doc/02-Installation.md#installing-requirements-)

**Comment by Al2Klimov** [comment]
Yes, I've rebased w/ master (#3214), but I'm afraid it didn't help.

**Comment by SMillerDev** [comment]
Any way I can help this PR get to production faster?

**Comment by Al2Klimov** [comment]
I'm afraid only @lippserd can do this... by just making [a simple decision](https://github.com/Icinga/icingaweb2/pull/3186#issuecomment-351960792).

**Comment by julianbrost** [comment]
FYI: The Archlinux package in the AUR now applies a patch based on this PR as Arch ships PHP 7.2. It's this pull request minus the changes in `modules/monitoring/application/views/helpers/PluginOutput.php` as the surrounding code got removed in v2.5.1. So this PR should receive a bit of testing there. In my setup it works fine so far.

**Comment by Al2Klimov** [comment]
Fine ... maybe [I was wrong](https://github.com/Icinga/icingaweb2/pull/3186#issuecomment-355996696) and @Thomas-Gelf can make that decision, too.

**Comment by Al2Klimov** [comment]
Why "a dedicated one" – and not this one? One nice day we just turn it on (disallow failing).

**Comment by Al2Klimov** [comment]
#3305, #3306, #3307

**Comment by Al2Klimov** [comment]
@Thomas-Gelf We don't have to allow Travis to fail – if you give green light on #3213

**Comment by Thomas-Gelf** [comment]
Perfect, merged all three of them, thank you. Same game again please for f334a9e and cf85e66. Also for f8a71b7, but there please extend the commit message and explain your motivation for this commit. Is our default error_reporting too low? Or too strict on PHP 7.2? What would fail (or not fail) without this change?

**Comment by Al2Klimov** [comment]
cf85e66? IMO we need this only here.

**Comment by Al2Klimov** [comment]
#3308, #3309

**Comment by Thomas-Gelf** [comment]
The only commit that will remain in this pull request is 2d5140a  - as it is the only one that fit's the topic. The others fix related issues. To me cf85e66 is clearly a bugfix. Parts of this have been triggered by 91b0e981, one line is older. I'm wondering why this didn't fail earlier, I would have expected also older PHP versions to warn on this.

Next, what was the motivation for 6fc90be (PhpCodeValidityTest)? It does obviously more than `php -l`, as it also loads the code. Also, lint-alike errors would usually be caught by phpcs.

And the last big beast is the session. That one requires special care. Some weird things in the original implementation derive from locking/performance reasons, as @lippserd pointed out. Not writing when nothing changed always makes sense, it saves a lot of I/O. Still, even with no change from time to time we need to force a write, as otherwise GC would kill our session.

Immediately closing the session on write is even more critical, as the default PHP behavior would have devastating effect on our performance. And what remains last:  modules like Nagvis are forced to do weird things with sessions. They shut one session down, start the other one, switch back - none of this works reliably when PHP headers are sent out early. This and related issues are what those INI settings should have taken care of.

To keep things short: there is too much duplicated code in PhpSession72. Please have a look at the write() method, just to name one example. There is one differing line, no need to clone the whole method. Please isolate the tasks causing trouble on 7.2 in the original file into dedicate methods and then override only them.

Thanks,
Thomas

**Comment by Al2Klimov** [comment]
Now as I've outsourced #3310, this works w/o #3213.

**Comment by Thomas-Gelf** [comment]
We're getting closer :-)

**Comment by Thomas-Gelf** [comment]
I cherry-picked c8930f3 to speed things up. Please move the fixes for the Session to a dedicated pull request and address above change requests. Thanks for the good work!

**Comment by Al2Klimov** [comment]
@Thomas-Gelf As long as Travis fails, please wait until I've rebased the other merged PRs.

**Comment by SMillerDev** [comment]
This is only waiting for the pr to drop 5.3 then?

**Comment by Al2Klimov** [comment]
@SMillerDev No, it isn't.

**Pull request merged**